src/aspire/abinitio/commonline_c3_c4.py

```diff
@@ -231,21 +237,270 @@ def _estimate_inplane_rotations(
     # Secondary Methods for computing outer product #
     #################################################
 
-    def _self_clmatrix_c3_c4(self, pf, order, max_shift, shift_step):
-        # return sclmatrix
-        pass
+    def _self_clmatrix_c3_c4(self):
+        """
+        Find the single pair of self-common-lines in each image assuming that the underlying
+        symmetry is C3 or C4.
+        """
+        pf = self.pf.copy()
+        n_img = self.n_img
+        n_theta = self.n_theta
+        max_shift_1d = np.ceil(2 * np.sqrt(2) * self.max_shift)
+        shift_step = self.shift_step
+        order = self.order
 
-    def _estimate_all_Riis_c3_c4(order, sclmatrix, n_theta):
-        # return Riis
-        pass
+        # The angle between self-common-lines is in the range [60, 180] for C3 symmetry
+        # and [90, 180] for C4 symmetry. Since antipodal lines are perfectly correlated
+        # we search for common lines in a smaller window.
+        # Note: matlab code used [60, 165] for order=3 and [90, 160] for order=4.
+        # We've increased this window to improve the MSE of our estimates.
+        if order == 3:
+            min_angle_diff = 60 * np.pi / 180
+            max_angle_diff = 175 * np.pi / 180
+        else:
+            min_angle_diff = 90 * np.pi / 180
+            max_angle_diff = 175 * np.pi / 180
+
+        # The self-common-lines matrix holds two indices per image that represent
+        # the two self common-lines in the image.
+        sclmatrix = np.zeros((n_img, 2))
+        corrs_stats = np.zeros(n_img)
+        shifts_stats = np.zeros(n_img)
+
+        # We create a mask associated with angle differences that fall in the
+        # range [min_angle_diff, max_angle_diff].
+        X, Y = np.meshgrid(range(n_theta), range(n_theta // 2))
+        diff = Y - X
+        unsigned_angle_diff = np.arccos(np.cos(diff * 2 * np.pi / n_theta))
```

Comment:
`x modulo π` is not quite the same. In this case, since we are looking at the angle difference between two common lines, we want to choose the smaller of the two angles formed. As an example, if two common lines form an angle of 190 degrees, then they also form an angle of 170 degrees. `np.arccos(np.cos(x))` gives us this "explementary" (had to look that up) angle. `x modulo π` would produce 10 degrees.

Ultimately, `np.arccos(np.cos(x))` gives us the unsigned value of the smaller of the two angles formed by a pair of common lines.

I will add a comment to explain what is happening here.

Reply:
Comment added.

Reply:
Then why not just compute that explicitly? Computing it through `arccos(cos(x))` just obfuscates things.

Reply:
Ok, I replaced `diff` with the below formulation and the comment "`diff` is the unsigned angle differences between all pairs of polar Fourier rays"

`diff = abs(theta_half - theta_full)`  
`diff[diff > 180] = 360 - diff[diff > 180]` 
`diff = np.deg2rad(diff)`

Then constructed the mask:
`good_diffs = (min_angle_diff < diff) & (diff < max_angle_diff)`

Maybe there is a clearer way to achieve this.